---
title: "ATAC RNA integration"
output: html_document
date: '2022-06-17'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Create scRNA-seq UMAP
```{r}
# Load the directory #-----#-----#-----#
setwd(your_working_directory) 
#**************************************#
#
#
# Load all libraries #-----#-----#-----#
library(Seurat); library(Matrix); library(R.utils)
#***************************************#
#
# Combine 3 non-mutant samples together.
fileNames = c( "Sample_WT-WERGFP", "Sample_WT-WERGFP_2", "Sample_WT-WERGFP_3")
data_sample = c()
for(i in fileNames){
  matrix = readMM(file = paste0(i, "/filtered_gene_bc_matrices/TAIR10/matrix.mtx"))
  rownames(matrix) = read.delim(file = paste0(i, "/filtered_gene_bc_matrices/TAIR10/genes.tsv"), header = FALSE, stringsAsFactors = FALSE)$V1
  colnames(matrix) = read.delim(file = paste0(i, "/filtered_gene_bc_matrices/TAIR10/barcodes.tsv"), header = FALSE, stringsAsFactors = FALSE)$V1
  data_sample <- c(data_sample, matrix)
}
ATH_data = cbind(data_sample[[1]], data_sample[[2]], data_sample[[3]]) # Combine data by columns
#
# Create seurat object for RNA-seq data
Rna <- CreateSeuratObject(ATH_data)
Rna <- FindVariableFeatures(object = Rna, selection.method = "vst", verbose = FALSE)
Rna <- NormalizeData(Rna)
Rna <- ScaleData(Rna, features = rownames(ATH_data))
Rna <- RunPCA(Rna, features = VariableFeatures(object = Rna) ,verbose = FALSE)
Rna <- FindNeighbors(Rna, reduction = "pca", dims = 1:31) 
Rna <- FindClusters(Rna, resolution = 0.55) 
Rna <- RunUMAP(Rna, reduction = "pca", dims = 1:31) 
DimPlot(Rna, reduction = "umap", label = TRUE) 

#***************************************#
cellType = read.csv(cell_type_file, row.names = 1)
Rna@meta.data$celltype = cellType$seurat_clusters
DimPlot(Rna, reduction = "umap", group.by = "celltype", label = TRUE, repel = TRUE)
head(Rna@meta.data, 5)

```

## Create scATAC-seq UMAP:
```{r}
par(mfrow = c(1,2))
Atac = readRDS(gunzip("GSM5281420_Dorrity_etal_scATAC_root.rds.gz", remove = FALSE))
DimPlot(Atac, reduction = "umap") +NoLegend()
DimPlot(Atac, group.by = "cell_type", label = TRUE, repel = TRUE, reduction = "umap") +NoLegend()
head(Atac@meta.data,5)
```
## 
```{r}
DefaultAssay(Atac) <- "geneactivity"
Atac <- NormalizeData(Atac)
Atac <- ScaleData(Atac, features = rownames(Atac))
#***************************************#
#
# Add technical-identify to metadata
Atac$tech = "ATAC"
Rna$tech = "RNA"

# Identify anchors
transferAnchors <- FindTransferAnchors(reference = Rna, query = Atac, features = rownames(Rna), 
                                       reference.assay = "RNA", query.assay = "geneactivity", 
                                       reduction = "cca")

celltype_pred <- TransferData(anchorset = transferAnchors, refdata = Rna$celltype, weight.reduction = Atac[["lsi"]], dims = 2:30)
Atac <- AddMetaData(Atac, metadata = celltype_pred) # Predict scATAC-seq cell type

```

## Co-embedding 2 datasets together
```{r}
imputation <- TransferData(anchorset = transferAnchors, 
                           refdata = GetAssayData(Rna, assay = "RNA", slot = "data")[rownames(Rna), ], 
                           weight.reduction = Atac[["lsi"]], dims = 2:30)
Atac[["RNA"]] <- imputation


# merge two data sets
Embedding <- merge(x = Rna, y = Atac)
Embedding <- ScaleData(Embedding, features = rownames(Rna), do.scale = FALSE)

# Use PCA and UMAP to visualize the co-embedding of two data sets on the same dimension
Embedding <- RunPCA(Embedding, features = rownames(Rna))
Embedding <- RunUMAP(Embedding, dims = 1:30)

# Export a pdf file of the image.
#pdf("061522_bookchapter.pdf", width = 7, height = 5)
DimPlot(Embedding, group.by = c("tech"))
#dev.off()

Embedding$celltype <- ifelse(!is.na(Embedding$celltype), Embedding$celltype, Embedding$predicted.id)
DimPlot(Embedding, group.by = "celltype", label = TRUE, repel = TRUE)
saveRDS(Embedding, "Integration_062022.rds")
```








